include/global/state.hpp

```diff
@@ -1089,6 +1121,112 @@ namespace NP {
+
+			// checks that `pred` is the only predecessor of `succ` that is not certainly finished
+			// or that `succ` is the only successor of all succ's predecessors (i.e., 
+			// the sum of all the successors of all predecessors of `succ` is equal to 1)
+			bool succ_ready_right_after_pred(const Job_index pred, const Job_index succ, const Successors& successors_of, const Predecessors& predecessors_of)
+			{
+				// if there is only one core then there is at most one job that is not certainly finished
+				if (num_cpus == 1)
+					return true;
+
+				// if `succ` has a single predecessors, then �t most one predecessor may not be finished
+				if (predecessors_of[succ].size() == 1)
+					return true;
+
+				for (const auto& p : predecessors_of[succ]) // check if all other predecessors of `succ` are certainly finished or that they have no other successors than `succ`
+				{
+					// if `succ` is not directly ready after `p` completes (i.e., there is a suspension delay between the completion of `p` and start of `succ`), 
+					// then `succ` is not certainly ready when `pred` finishes 
+					if (p.second.max() > 0)
+						return false;
+
+					Job_index j_idx = p.first->get_job_index();
+					// if `p` was not dispatched yet, then `succ` cannot be ready
+					if (!scheduled_jobs.contains(j_idx))
+						return false;
+
+					//  if `p` has a single successor, then we disregard it
+					if (successors_of[j_idx].size() == 1)
+						continue;
+
+					if (j_idx != pred)
+					{
+						// If at least one successor of `p` has already been dispatched, then `p` must have finished already.
+						bool is_finished = false;
+						for (const auto& succ_of_pred : successors_of[j_idx])
+						{
+							if (scheduled_jobs.contains(succ_of_pred.first->get_job_index())) {
+								is_finished = true;
+								break;
+							}
+						}
+						if (is_finished == false)
+							return false;
+					}
+				}
+				return true;
+			}
+
+			void update_ready_successor_jobs_prio(const Schedule_node& from,
+				const Job<Time>& j, const Successors& successors_of,
+				const Predecessors& predecessors_of,
+				const Job_set& scheduled_jobs)
+			{
+				ready_successor_jobs_prio.reserve(from.ready_successor_jobs_prio.size() + 1);
+				Job_index j_idx = j.get_job_index();
+
+				bool job_to_insert = false;
+				Priority max_prio = Time_model::constants<Time>::infinity();
+				Job_index job_id;
+
+				// find the highest priority successor of j that that will be ready as soon as j finishes its execution
+				for (const auto& s : successors_of[j_idx]) {
+					Job_index succ_id = s.first->get_job_index();
+
+					// if `s` was not dispatched yet, can execute on a single core, is ready right after `j` finishes, and has no other predecessor than `j` or all other predecessors certainly finished
+					if (!scheduled_jobs.contains(succ_id)
+						&& s.first->get_min_parallelism() == 1 
+						&& s.second.max() == 0 && s.first->latest_arrival() <= (j.earliest_arrival()+j.least_exec_time())
```

Comment:
In a later PR, we can probably do something tighter than `j.earliest_arrival()`, by remembering the start time interval of `j`.

Reply:
Correct. By moving the logic to Schedule_state, we can use the earliest finish time of `j` instead of its (earliest arrival + bcet). I will push that tomorrow hopefully